fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
+                    bucket.current_offset,
+                    param.grad.numel(),
                 )
 
-            # Directly reduce the other grads
-            for p in params[i_bucketed:]:
-                p.grad = cast(Tensor, p.grad)
-                if p.grad.requires_grad:
-                    raise RuntimeError("DistributedDataParallel only works with gradients that don't require grad")
+                if bucket.full():
 
-                p.grad.div_(world_size)
-                requests.append(dist.reduce(tensor=p.grad, dst=global_rank, group=group, async_op=True))  # type: ignore
+                    def unwrap() -> None:
+                        for flat in bucket.params:
+                            if dst_rank != self.global_rank:
+                                # this rank is not the owner, release the grad
+                                flat.param.grad = None
+                            else:
+                                # this rank is the owner, unroll the results
+                                assert flat.param.grad is not None
 
-        # Unroll the initial packed small gradients, as soon as possible
-        for future, rank in bucket_requests:
-            future.wait()
+                                flat.param.grad.data.copy_(
+                                    bucket.buffer[flat.start : flat.stop].view_as(flat.param.data), non_blocking=True
+                                )
 
-            if rank == self_rank:
-                i_bucketed = 0  # the number of tensors packed in the buffer
-                offset = 0
-                params = per_rank_params[rank]
-                buffer = buffers[rank]
+                        bucket.reset()
 
-                while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
-                    end = offset + params[i_bucketed].numel()
-                    params[i_bucketed].grad.data.copy_(buffer[offset:end].view_as(params[i_bucketed]))  # type: ignore
-                    offset = end
-                    i_bucketed += 1
+                    bucket.buffer /= self.world_size
 
-        # Make sure that we're done with this device before moving on and cleaning the unused params
-        _ = list(map(lambda x: x.wait(), requests))
+                    optimizer.work_handles.append(
+                        Workhandle(
+                            handle=dist.reduce(
+                                tensor=bucket.buffer, dst=dst_rank, group=self.process_group, async_op=True,
+                            ),
+                            callback=unwrap,
+                        )
+                    )
+
+                    # If all the reduce operations have been called, add the gatekeeper
+                    if len(optimizer.work_handles) == optimizer._max_work_handles:
+                        gatekeeper()
 
-    def _sync_buffers(self) -> None:
+        return reduce_bucket if should_bucket else reduce_direct
```

Comment:
Might be cleaner to define reduce_direct and reduce_bucket as instance methods and call the appropriate one depending on should_bucket?

Reply:
it's a little bit more complicated than that, because they will be called as hooks, and in that case the only allowed parameter is 'grad' (so I cannot pass the destination rank for instance, [see](https://pytorch.org/docs/stable/autograd.html?highlight=register_hook#torch.Tensor.register_hook) ). That's why they are closures here and saved in the pseudo future (work handle list), because they capture these extra parameters, I don't think that it's doable as a method. One thing to remember is that I don't even know in which order they will be called, so there's no extra context that I use to guess these. It should be greatly simplified with proper futures, this is being worked on I think (cc @mrshenli to make sure I'm not completely wrong on that)

Reply:
hmm, actually I think that's doable but some parts would be more complicated (find out where the hook is called from given the grad), especially since said grads are typically dropped if this rank does not own them (so I cannot have a static LUT)